lib/grammers-client/src/types/photo_sizes.rs

```diff
@@ -0,0 +1,284 @@
+use std::path::Path;
+
+use tokio::fs;
+use tokio::io::AsyncWriteExt;
+
+use grammers_tl_types as tl;
+
+use crate::ClientHandle;
+
+pub enum PhotoSize {
+    Empty(SizeEmpty),
+    Size(Size),
+    Cached(CachedSize),
+    Stripped(StrippedSize),
+    Progressive(ProgressiveSize),
+    Path(PathSize),
+}
```

Comment:
I don't know if it's desirable to expose this much information. I think a `struct PhotoSize` with `impl` for `cached() -> bool`, `dim()` for dimensions, `width()`, `is_vector()` would be a lot more useful. What do you think?

In addition, I believe Telegram borrowed the `photo_type` string format from [VK Photo sizes format](https://vk.com/dev/photo_sizes). Maybe the `enum Size` for `Small`, `Medium`, etc. would be better (accidentally lost this comment as a part of the review and had to type it again).

Reply:
Well, my position is that good abstraction over the types should not hide the raw information. I mean, get more helper helpers (dim, width, is_vector) is great, but it's not okay if the user won't be able to get `StrippedSize` type. For example, maybe the user doesn't want to build a full featured application, but only a "super stripped photo sizes extractor".

But maybe this crate has a different purpose. If this is not about "you can build any type of application", but "you have a nice API to build applications fast and simple and for more complicated cases, please use raw calls and types", then we can provide only such methods.

Reply:
> If this is not about "you can build any type of application", but "you have a nice API to build applications fast and simple and for more complicated cases, please use raw calls and types", then we can provide only such methods.

Yes, that's it. The Rust community values semantic versioning a lot, and with Telegram changing things so often with new layers, most of the time it would be a breaking change. It would be silly to bump the major version of grammers on every new layer, because it technically is a breaking change. We'd end up with huge versions in no time, so semver becomes pointless.

`grammers-client` is a high-level client which you're not forced to use; you could rely on just `grammers-mtsender` and implement your abstractions your own way. And there will be a way to provide access to the raw types. But the public API should be simple, and avoid introducing breaking changes as much as possible.

Note I'm not saying we should hide `StrippedSize` away, we should offer a convenient way to access this information. `is_cached` would do well, as that's the main purpose of the type.

Reply:
Oh, thank you for the explanation. I didn't know about the issue with a frequent telegram schema updates.

Reply:
You can view all the layers and their changes at https://diff.telethon.dev/. Generally, there's a few every year (and sometimes they change the layer without incrementing the layer number). Providing a friendlier abstractions will make the client more "rusty" and also more resistant to changes.